### PR TITLE
Bitrise xcui plugin - initial commit

### PR DESCRIPTION
**Pull request by irishabh96**

### CODE REVIEW

File: bitrise.yml [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+default_step_lib_source: https://github.com/bitrise-io/bitrise-steplib.git
+
+app:
+  envs:
+  # An example secret param, define it (A_SECRET_PARAM) in .bitrise.secrets.yml
+  - A_SECRET_PARAM: $A_SECRET_PARAM
+  # If you want to share this step into a StepLib
+  - BITRISE_STEP_ID: xcui-browserstack-official
+  - BITRISE_STEP_VERSION: "0.0.1"
+  - BITRISE_STEP_GIT_CLONE_URL: https://github.com/irishabh96/browserstack-bitrise-xcui-step.git
```
Comment (deepwired) [inline bitrise.yml]:
```suggestion
  - BITRISE_STEP_GIT_CLONE_URL: https://github.com/browserstack/browserstack-bitrise-xcui-step.git
```

File: util_fns.go [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+
+	"github.com/olekukonko/tablewriter"
+)
+
+func getDevices() ([]string, error) {
+	var devices []string
+
+	devices_input := os.Getenv("devices_list")
+
+	if devices_input == "" {
```
Comment (deepwired) [inline util_fns.go:23]: Do we need more validations here?
Do we want to try to type cast to a list of strings? etc..

**Review by deepwired — CHANGES_REQUESTED** [review]

File: util_fns.go (hunk 2 above)
Comment (irishabh96) [reply util_fns.go:23]: By default every input is a string in Bitrise, and any list is a line break. So the function (scanner.scan() & scanner.text()) which I'm using to get the text from input does this validation by default. Reference - [Link](https://medium.com/golangspec/in-depth-introduction-to-bufio-scanner-in-golang-55483bb689b4)

File: main.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+)
+
+func main() {
+	// Exit after 2hours 30 mins
+	time.AfterFunc(150*time.Minute, func() { failf("Session Timed Out") })
+
+	log.Print("Starting the build on BrowserStack App Automate")
+
+	username := os.Getenv("browserstack_username")
+	access_key := os.Getenv("browserstack_accesskey")
```
Comment (avinash-bharti) [inline main.go:19]: Should we move these into contant file and consume the constant here i.e BROWSERSTACK_USERNAME := os.Getenv("browserstack_username")

File: main.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+	ios_app := os.Getenv("app_ipa_path")
+	test_suite_path := os.Getenv("xcui_test_suite")
```
Comment (avinash-bharti) [inline main.go:21]: Should we move these into constant file and consume the constant here i.e BROWSERSTACK_USERNAME := os.Getenv("browserstack_username")

File: util_fns.go [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+func locateTestRunnerFileAndZip(test_suite_location string) error {
+	split_test_suite_path := strings.Split(test_suite_location, "/")
+	get_file_name := split_test_suite_path[len(split_test_suite_path)-1]
+
+	test_runner_app_path := ""
+
+	check_file_extension := strings.Split(get_file_name, ".")
+
+	// Checking 2 conditions here
+	// 1. test_suite_location - is this a the runner app
```
Comment (avinash-bharti) [inline util_fns.go]: minor typo

File: services.go [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+	if build_id == "" {
+		return "", errors.New(fmt.Sprintf(FETCH_BUILD_STATUS_ERROR, "invalid build_id"))
+	}
+
+	if waitForBuild {
+		log.Println("Waiting for results")
+	}
+
+	// ticker can't have negative value
+	var POOLING_INTERVAL int = 1000
```
Comment (avinash-bharti) [inline services.go]: Please rename all `POOLING` to POLLING.

File: main.go (hunk 4 above)
Comment (irishabh96) [reply main.go:21]: Discussed over call, marking as resolved.

**Review by deepwired — DISMISSED** [review]
LGTM

**Review by avinash-bharti — DISMISSED** [review]
LGTM

**Review by deepwired — APPROVED** [review]

**Pull request merged**